File: lib/metrics.ts
import type { TaggedEvent, WeeklyMetrics, SectorMetrics } from './models'
import { normalizeSectors, type CanonicalSector } from './sectors'

function computeWsi(events: TaggedEvent[]): number {
  const valid = events.filter(e => e.sentiment_score !== null && e.conviction !== null)
  if (!valid.length) return 0
  const num = valid.reduce((s, e) => s + e.sentiment_score! * e.conviction! * e.channel_weight, 0)
  const den = valid.reduce((s, e) => s + e.conviction! * e.channel_weight, 0)
  return den ? num / den : 0
}

function computeDivergence(events: TaggedEvent[]): number {
  const scores = events.map(e => e.sentiment_score).filter((s): s is number => s !== null)
  if (scores.length < 2) return 0
  const mean = scores.reduce((a, b) => a + b, 0) / scores.length
  const variance = scores.reduce((s, x) => s + (x - mean) ** 2, 0) / scores.length
  return Math.sqrt(variance)
}

function checkAlerts(velocity: number, divergence: number): string[] {
  const alerts: string[] = []
  if (Math.abs(velocity) > 0.3) {
    const dir = velocity > 0 ? '상승' : '하락'
    alerts.push(`VELOCITY_ALERT: ${velocity.toFixed(3)} — 주간 톤 급변(${dir})`)
  }
  if (divergence > 0.5) {
    alerts.push(`DIVERGENCE_ALERT: ${divergence.toFixed(3)} — 의견 양극화 (변곡점 가능성)`)
  }
  return alerts
}

// Groups tagged events by canonical sector and computes per-sector metrics.
// eventsNow = this week, eventsPrev = previous week (for velocity).
export function buildAllSectorMetrics(
  eventsNow: TaggedEvent[],
  eventsPrev: TaggedEvent[],
): Record<CanonicalSector, SectorMetrics> {
  const groupNow = groupBySector(eventsNow)
  const groupPrev = groupBySector(eventsPrev)

  const result = {} as Record<CanonicalSector, SectorMetrics>

  const sectors = new Set<CanonicalSector>([
    ...Object.keys(groupNow),
    ...Object.keys(groupPrev),
  ] as CanonicalSector[])

  for (const sector of sectors) {
    const now = groupNow[sector] ?? []
    const prev = groupPrev[sector] ?? []

    const wsiNow = computeWsi(now)
    const wsiPrev = computeWsi(prev)
    const velocity = wsiNow - wsiPrev
    const divergence = computeDivergence(now)

    const topTickers = getTopTickers(now)

    result[sector] = {
      wsi: Math.round(wsiNow * 10000) / 10000,
      velocity: Math.round(velocity * 10000) / 10000,
      divergence: Math.round(divergence * 10000) / 10000,
      event_count: now.length,
      top_tickers: topTickers,
      alerts: checkAlerts(velocity, divergence),
    }
  }

  return result
}

function groupBySector(events: TaggedEvent[]): Partial<Record<CanonicalSector, TaggedEvent[]>> {
  const groups: Partial<Record<CanonicalSector, TaggedEvent[]>> = {}
  for (const event of events) {
    const canonical = normalizeSectors(event.sectors)
    for (const sector of canonical) {
      if (!groups[sector]) groups[sector] = []
      groups[sector]!.push(event)
    }
  }
  return groups
}

function getTopTickers(events: TaggedEvent[]): string[] {
  const counts: Record<string, number> = {}
  for (const e of events) {
    for (const t of e.tickers) {
      counts[t] = (counts[t] ?? 0) + 1
    }
  }
  return Object.entries(counts)
    .sort((a, b) => b[1] - a[1])
    .slice(0, 3)
    .map(([t]) => t)
}

export function buildWeeklyMetrics(
  ticker: string,
  weekStart: string,
  eventsNow: TaggedEvent[],
  eventsPrev: TaggedEvent[],
  provider = '',
): WeeklyMetrics {
  const wsiNow = computeWsi(eventsNow)
  const wsiPrev = computeWsi(eventsPrev)
  const velocity = wsiNow - wsiPrev
  const divergence = computeDivergence(eventsNow)
  return {
    ticker,
    week_start: weekStart,
    wsi: Math.round(wsiNow * 10000) / 10000,
    velocity: Math.round(velocity * 10000) / 10000,
    divergence: Math.round(divergence * 10000) / 10000,
    event_count: eventsNow.length,
    alerts: checkAlerts(velocity, divergence),
    provider,
  }
}

File: src/metrics.py
import statistics
from .models import TaggedEvent, WeeklyMetrics


def compute_wsi(events: list[TaggedEvent]) -> float:
    """Weighted Sentiment Index: Σ(sentiment × conviction × weight) / Σ(conviction × weight)"""
    valid = [e for e in events if e.sentiment_score is not None and e.conviction is not None]
    if not valid:
        return 0.0
    numerator = sum(e.sentiment_score * e.conviction * e.channel_weight for e in valid)
    denominator = sum(e.conviction * e.channel_weight for e in valid)
    return numerator / denominator if denominator else 0.0


def compute_velocity(wsi_now: float, wsi_prev: float) -> float:
    return wsi_now - wsi_prev


def compute_divergence(events: list[TaggedEvent]) -> float:
    """Std dev of sentiment scores across channels — high = opinion split = inflection signal."""
    scores = [e.sentiment_score for e in events if e.sentiment_score is not None]
    return statistics.stdev(scores) if len(scores) > 1 else 0.0


def check_alerts(velocity: float, divergence: float,
                 claim_cluster_shift: float = 0.0,
                 new_claims_channel_count: int = 0) -> list[str]:
    alerts = []
    if abs(velocity) > 0.3:
        direction = "상승" if velocity > 0 else "하락"
        alerts.append(f"VELOCITY_ALERT: {velocity:+.3f} — 주간 톤 급변({direction})")
    if divergence > 0.5:
        alerts.append(f"DIVERGENCE_ALERT: {divergence:.3f} — 의견 양극화 (변곡점 가능성)")
    if claim_cluster_shift > 0.4:
        alerts.append(f"CLUSTER_SHIFT_ALERT: {claim_cluster_shift:.0%} — narrative 변화")
    if new_claims_channel_count >= 5:
        alerts.append(f"COLLECTIVE_RECOGNITION: {new_claims_channel_count}개 채널 동시 신규 claim 등장")
    return alerts


def build_weekly_metrics(ticker: str, week_start: str,
                          events_now: list[TaggedEvent],
                          events_prev: list[TaggedEvent],
                          provider: str = "") -> WeeklyMetrics:
    wsi_now = compute_wsi(events_now)
    wsi_prev = compute_wsi(events_prev)
    velocity = compute_velocity(wsi_now, wsi_prev)
    divergence = compute_divergence(events_now)
    alerts = check_alerts(velocity, divergence)
    return WeeklyMetrics(
        ticker=ticker,
        week_start=week_start,
        wsi=round(wsi_now, 4),
        velocity=round(velocity, 4),
        divergence=round(divergence, 4),
        event_count=len(events_now),
        alerts=alerts,
        provider=provider,
    )
